agentos/component.py

```diff
@@ -223,3 +141,111 @@ def _parse_components(cls, components_spec: Dict, repos: Dict) -> Dict:
                 component.add_dependency(dependency, alias=alias)
 
         return components
+
+    def run(self, fn_name: str, params: ParameterSet = None):
+        params = params if params else ParameterSet()
+        with self._track_run(fn_name, params):
+            return self._untracked_run(fn_name, params)
+
+    def _untracked_run(
+        self, fn_name: str, params: ParameterSet, instance=None
+    ):
+        instance = instance if instance else self.get_instance(params=params)
+        fn = getattr(instance, fn_name)
+        assert fn is not None, f"{instance} has no attr {fn_name}"
+        fn_params = params.get(self.name, fn_name)
+        print(f"Calling {self.name}.{fn_name}(**{fn_params})")
+        result = fn(**fn_params)
+        return result
+
+    def add_dependency(self, component: "Component", alias: str = None):
+        if type(component) is not type(self):
+            raise Exception("add_dependency() must be passed a Component")
+        if alias is None:
+            alias = component.name
+        self._dependencies[alias] = component
+
+    def get_instance(self, params: ParameterSet) -> None:
+        instantiated = {}
+        return self._get_instance(params, instantiated)
+
+    def _get_instance(self, params: ParameterSet, instantiated: dict) -> T:
+        if self.name in instantiated:
+            return instantiated[self.name]
+        save_init = self._managed_cls.__init__
+        self._managed_cls.__init__ = lambda self: None
+        instance = self._managed_cls()
+        for dep_alias, dep_component in self._dependencies.items():
+            print(f"Adding {dep_alias} to {self.name}")
+            dep_instance = dep_component._get_instance(
+                params=params, instantiated=instantiated
+            )
+            setattr(instance, dep_alias, dep_instance)
+        setattr(instance, self._dunder_name, self)
+        self._managed_cls.__init__ = save_init
+        self._untracked_run("__init__", params=params, instance=instance)
+        instantiated[self.name] = instance
+        return instance
+
+    def _log_params(self, params: ParameterSet) -> None:
+        log_data_as_yaml_artifact("parameters.yaml", params.to_dict())
```

Comment:
Can we call it `component_parameter_set.yaml` or just `parameter_set.yaml` instead of `parameter.yaml`?